Feature/plot api

bradyzp opened the pull request:
Plot API Feature Branch
===============

Resolves #37 
Resolves #54 
Closes #36 

**Overview**
This is a major re-write of the core plotting wrappers/helpers/integration-framework (hopefully for the last time :) ). The primary modules of interest here are located in the dgp/gui/plotting/ package:

- backends.py 
- plotters.py 
- helpers.py

The plotting API for use within DGP has been re-written and cleaned up significantly, with various enhancements and improvements made. The new API is based exclusively on the pyqtgraph library due to its interactive performance.

**Features/Improvements**

- GridPlotWidget and its derivatives now directly subclass GraphicsView, and as such can be directly added to any QWidget/Layout
- Custom PlotItem (DgpPlotItem) supports (optional) left/right-y axis for plotting multiple data sets with varying scales on the same plot surface.
- Custom PolyAxis axis formatter allows selective formatting of axis ticks (can switch between scalar and date/time format) - this is useful for example in the Transform work-space as the user may switch between time/longitude/latitude for the x-axis data index.
- Fixed/improved labels for data segment selection areas
- Add custom plot controls widget (via right-click context menu) to plots, replacing the default pyqtgraph controls (many of which would cause RuntimeErrors)
- PlotTab and TransformTab workspaces updated to utilize new plotting API. PlotTab makes use of dual y-axis plots, and automatically plots to the left/right axis based on a cutoff value.
- Improved control of Data Segment selection in the LineSelectPlot; DataSegments/Controllers support referencing their plotted representations and can effect mutations on any active plotted segments (and vice versa)
- Improved/new documentation for the plotting API. Added plotting sub-module page to sphinx build.
- Minor cosmetic updates to TransformTab, experimental feature added to extract transform graph definition from the source-code to enable display/inspection within the application.

**ToDo** (Future Goals)
Add an export mechanism for exporting/generating static images from plotted data (possibly use Matplotlib to create pretty non-interactive plots). Note that I have disabled/removed the export functionality that pyqtgraph provides as many of its functions don't work correctly or cause runtime errors.

Save/restore plot state, enable user to select to track the plot state, so that data items can be re-plotted automatically when opening/closing a tab.

Enable fine-tuned control and/or creation of LinearSegments on the LineSelectPlot, allowing the user to adjust the left/right bounds via a dialog, and possibly a dialog which would enable the input of specific time ranges to create a series of segments.

bradyzp committed:
Add test-driven re-write of the PyQtGridPlotWidget

Improved API for GridPlotWidget class, removes a layer of indirection where the plot widget was encapsulated within the outer class.
Tests added for most API calls in the new GridPlotWidget class.
Added new PolyAxis class which will replace DateAxis, this Axis can be switched to display DateTime labels, or scalar labels.

bradyzp committed:
Update PolyAxis date tickStrings function.

Updated and added to the tickStrings function for generating tick strings on a Date/Time axis.
Added testing to verify proper tick formats are generated.

bradyzp committed:
ENH: Add second y-axis scale to plots.

Add ability to enable secondary (right) y-axis scale to a plot within a
GridPlotWidget. This enables plotting of two different Series with
differing magnitudes on the same plot.
Added/updated plot tests.
Update LinearFlightRegion context menu actions to work with new
LineSelectPlot API.

bradyzp committed:
Re-implement LineSelectPlot API on new backend.

Add tests/API functionality for new LineSelectPlot based on new GridPlotWidget.
Add APP import in pytest conftest.py to remove the need to manually import the Qt APP object for tests requiring Qt support.

DOC: Add gui/plotting documentation pages to sphinx build.

bradyzp committed:
ENH: Minimally functional LineSelectPlot with Date axis formatter.

Fix date formatting on PolyAxis. Date axis will now also show the full y/m/d for the first tick displayed on the plot.
ToDo: PolyAxis dateTickStrings tests need to be re-done. There are probably still some tweaks to make to the date format strings.

bradyzp committed:
Refactor DataSegment to natively use pandas.Timestamp

Refactor DataSegment/Controller and the LineSelectPlot to natively use pandas.Timestamp, instead of converting back and forth between float (timestamps), Python datetime objects, and pandas Timestamp objects.
The DataSegment model provides getter/setters that handle Timestamp input appropriately, and in future the pandas Timestamp object will be supported by the project serialization interface, meaning that Timestamps can be direct attributes of model objects.

bradyzp committed:
Add enumerations to replace string parameter matching.

Added the enumerations AxisFormatter and StateAction which define types for specifying the string formatter of a plot axis (DATETIME or SCALAR) and the StateAction enum which is used by the LineUpdate tuple to notify the interface when
a line segment has been created/updated/deleted.
Plotter backends also adds an Axis enum which simply defines the Left/Right axis pairs (for use when a plot has a twin-y axis).

bradyzp committed:
Convert TransformPlot to new Plot API. Cleanup old plot API.

Convert TransformPlot class to use new GridPlotWidget
Clean up deprecated code (PyQtGridPlotWidget and DateAxis)
Update default parameters for LineSelectPlot to create multi-y axis.

bradyzp committed:
Update segment label handling and decouple from parent plot object.

LinearFlightRegion now exposes signals which can be listened to by the parent to trigger an update of its siblings labels instead of directly referencing the untyped parent object.
A new signal is also added for the LFR to request a deletion (as its siblings must also be deleted by the parent [plot] object).

bradyzp committed:
Transform Workspace improvements.

Added mode selection to Transform Tab - this allows the user to plot transformation results as a single line, or when toggled to break the resultant series into segments defined by the user in the line selection plot.
Add experimental method to view the source-code for the current TransformGraph function. This extracts the __init__ source code from the graph sub-class, and displays it within a QTextEdit. If the Pygments library is installed the source will also be highlighted.

bradyzp committed:
Tests: Qt/pytest Error handling improvements

Changed pytest run command to disable stdout/stderr capture on travis CI build. This was an issue when the test encounters a critical error (causing the Python interpreter to crash), resulting in the loss of any useful capture information.
Added new Qt error message handler in the pytest conftest file to capture and print to stderr any QFatal/QCritical calls.
Added pytest-faulthandler module to test-requirements which will print out a call trace in the event of even a fatal Python error.

Fix enumeration references after refactoring.

bradyzp committed:
Refactor AxisFormatter and LineraFlightRegion

Move AxisFormatter enum into plotting/backends
Rename LinearFlightRegion to LinearSegment
Change slot y_changed to y_rng_changed in LinearSegment

bradyzp committed:
Add custom DgpPlotItem with enhanced context menu.

Create DgpPlotItem as a derivative of pyqtgraph's PlotItem to override the default 'Plot Options' context menu provided by it. The default menu provides many options which are either irrelevant, or cause RuntimeErrors.
DgpPlotItem also encapsulates the process of creating a twin plot for the right y-axis, and also enables control of the twin plot (if it is enabled) via the new context menu.
Move twin y-axis plot creation into DgpPlotItem, 'right' plot can now be enabled by an init parameter.

bradyzp committed:
Refactor GridPlotWidget to use weak refs. Add docstrings.

Refactor GridPlotWidget to use WeakValueDictionaries to maintain weak references to plotted series and plot data items. This obviates the need to explicitly delete keys from both dictionaries when a curve is removed from a plot.
Updated types and docstrings for various methods in plotting backends.

bradyzp committed:
Add tooltips to plot controls. Fix 'View All' action behavior on multi-Y plots.

bradyzp committed:
Move LineUpdate named tuple into plotting/helpers.

bradyzp committed:
Add LinearSegmentGroup, refactor Line Select Plot logic.

Add LinearSegmentGroup class to logically group LinearSegments, and to handle updates/mutations within the group.

bradyzp committed:
Implement DataSegment reference tracking.

Add ability for DataSegmentController to track references to representations of itself painted in a plot. This enables actions on data-segments to be performed in a bi-directional manner, with the ability to mutate and propagate changes from both the actual plot surface, and via the DataSegmentController representation in the project tree view.
Add Delete action to DataSegmentController
Remove BaseController concrete class from DataSegmentController inheritance, use IBaseController instead.

bradyzp committed:
Add public utility methods to backends/plotters

Add autorange method to GridPlotWidget to allow external callers to auto-range every plot within the widget.
Refactor LineSelectPlot's selection_mode property to allow for signal connection.

bradyzp committed:
Update docstrings, finish sphinx plotting documentation page

Update, add, and edit docstrings on various plotting API methods/classes.
Fix various class/function links in RST documentation pages.
Add missing classes/attributes to sphinx plotting page.

cbertinato commented on dgp/gui/workspaces/PlotTab.py:
```diff
@@ -67,9 +76,10 @@ def _setup_ui(self):
-        qvbl_plot_layout.addWidget(self.plot.widget)
+        # self.plot.widget.setSizePolicy(QSizePolicy(QSizePolicy.Expanding,
+        #                                            QSizePolicy.Expanding))
+        # qvbl_plot_layout.addWidget(self.plot.widget)
```
Still need this code?

cbertinato commented on dgp/gui/workspaces/PlotTab.py:
```diff
@@ -79,37 +89,45 @@ def _setup_ui(self):
+            axis = Axis.LEFT
+        self._plot.add_series(item.data(Qt.UserRole), row, axis=axis)
+        # plot = self._plot.get_plot(row=row)
+        # plot.autoRange(items=plot.curves)
```
Same.

cbertinato commented on dgp/gui/workspaces/PlotTab.py:
```diff
@@ -79,37 +89,45 @@ def _setup_ui(self):
 
     def _toggle_selection(self, state: bool):
-        self.plot.selection_mode = state
+        self._plot.selection_mode = state
```
This throws an `AttributeError` because there is no setter selection_mode in `LineSelectPlot`. I suppose you meant to use `set_select_mode` here.  I found this when trying to toggle into selection mode in the UI, but I see that there is a unit test for this method and it uses `set_select_mode`.

cbertinato commented on dgp/gui/workspaces/TransformTab.py:
```diff
@@ -105,101 +126,113 @@ def _segments(self) -> List[DataSegmentController]:
+        the way the graph is defined in the source. The current method gets the
+        __init__ source code for the TransformGraph descendant then searches for
+        the string index of 'self.transform_graph', and takes from the first '{'
+        until the first '}'.
```
We will soon be able to overhaul this soon with the TransformGraph revamp, but it's good to see what you are doing here so I can provide a better way of doing this. Let's talk online more about what you're thinking here in terms of functionality.

bradyzp committed:
Add integrated toolbar factory method to plots.

The base plotter class (GridPlotWidget) and its descendants now provide a get_toolbar() method which constructs a QToolBar that can be added to a UI layout.
The toolbar provides actions that are common to all plots, for example allowing the user to auto-scale the view to fit the data, hide/show grid-lines etc.
The toolbar can be extended by descendant classes to add other task specific actions, e.g. the LinePlotWidget adds a toolbar action to toggle its line selection mode.

Note that some of the icon definitions (enumerations) added here are placeholders for icons that are added in a separate future feature branch.

bradyzp commented:
Latest commit adds new toolbar factory methods to the plotting base class. The PlotTab and TransformTab have been refactored slightly to make use of the new toolbar instead of defining its own plot-control actions.

This should resolve the attribute error encountered due to the missing selection_mode property. And removes the dead-code that was noted above.

*Note* The icons for the new toolbars have not been added to the project in this branch and so will show up simply as buttons with text. This will be fixed in an upcoming feature branch as I didn't want to pollute this branch with the addition/deletion of a bunch of icon/resource files.
